### PR TITLE
run upgrades twice

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

File: bats/fb-katello-proxy.bats [hunk 1]
```diff
@@ -23,6 +23,8 @@ setup() {
 }
 
 @test "sync proxy" {
+  # workaround for https://projects.theforeman.org/issues/32275
+  hammer setting set --name check_services_before_actions --value false
```
Comment (ekohl) [inline bats/fb-katello-proxy.bats:27]: Should this be behind a version check?

**Review by ekohl — COMMENTED** [review]
From a conceptual point of view I like this approach.

File: bats/fb-katello-proxy.bats (hunk 1 above)
Comment (evgeni) [reply bats/fb-katello-proxy.bats:27]: honestly, it shouldn't be here, at all. I hope we can just fix that for 4.0 GA :)

File: pipelines/upgrade/08-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+---
+- name: run tests
+  hosts:
+    - "{{ forklift_server_name }}"
+  become: yes
+  vars_files:
+    - ../vars/upgrade_base.yml
+    - ../vars/repos_staging.yml
+  vars:
+    bats_output_dir: "/root/bats_results_upgrade1"
```
Comment (ehelms) [inline pipelines/upgrade/08-tests.yaml:10]: What about `bats_results_proxy_n-1_upgrade` ? Or some such naming to make it easier when looking through the data to remind ourselves what each means.

**Comment by evgeni** [comment]
while unrelated to this change, I am getting the following error quite often on a katello 3.18 box:

```
    # Error: RPM1004: Error retrieving metadata: Error 'Proxy Error' for https://pipe-up-katello-3-18-centos7.kangae.example.com/pulp/repos/Test_Organizationupgrade1/Library/Component_1/custom/Test_Product/modules-rpms/repodata/ee66d94a8356524e705464ff75b6a3b98d7a362c060ff34417a9981d9fab5851-modules.yaml.
```

which is rather annoying :(

**Comment by ehelms** [comment]
I think that is a known issue -- let's double check with @jlsherrill

**Comment by evgeni** [comment]
Yeah, it's just now that we run the tests twice, we have a 2× chance to hit it :(

**Comment by jlsherrill** [comment]
I believe it was fixed by https://github.com/theforeman/puppet-pulpcore/commit/cb40e2805d858459e2820e2db05babba39042b68

which we seem to be missing from 3.18 :(

**Comment by ehelms** [comment]
> I believe it was fixed by [theforeman/puppet-pulpcore@cb40e28](https://github.com/theforeman/puppet-pulpcore/commit/cb40e2805d858459e2820e2db05babba39042b68)
> 
> which we seem to be missing from 3.18 :(

Guess we either need to backport it or skip this test when it's 3.18.

**Comment by evgeni** [comment]
The failing test is "sync a capsule" which I'd rather not skip ;)

**Comment by ekohl** [comment]
I've marked https://projects.theforeman.org/issues/31815 as a 2.3.4 cherry pick (and corrected the fixed in version).

**Comment by evgeni** [comment]
now that the pulp2 ping issue is fixed, I am facing https://projects.theforeman.org/issues/31401 again

**Comment by evgeni** [comment]
This is blocked by the following issues:
* Katello:
  * [#32426 - katello should pass Docker-Content-Digest from pulp3 to the container client](https://projects.theforeman.org/issues/32426) -- merged in nightly only
* Pulp:
  * [#8629 - Schema converter returns digest value without digest algoritm](https://pulp.plan.io/issues/8629) - backported to `pulp_container` 2.1.2 in [#8655](https://pulp.plan.io/issues/8655)
  * [#8631 - schema conversion adds an empty blob in the manifest but it is not present on the fs](https://pulp.plan.io/issues/8631) - backported to `pulp_container` 2.1.2 in [#8654](https://pulp.plan.io/issues/8654)

@jlsherrill do I see it right, once we have the Katello fix backported to 4.0(.1) and updated pulp_container to 2.1.2 we're good here?

**Comment by jlsherrill** [comment]
@evgeni i believe that sounds right.  I marked https://projects.theforeman.org/issues/32426 for 4.0.1

**Comment by ehelms** [comment]
Katello 4.0.1 is now out with the fix. Pulp 3.9 repository has pulp_container 2.2.1 (http://yum.theforeman.org/pulpcore/3.9/el7/x86_64/python3-pulp-container-2.2.1-1.el7.noarch.rpm) does that mean all issues are addressed?

**Comment by evgeni** [comment]
Argh, I'm stupid. 4.0 uses pulp 3.9, and that has not received the container fixes.

They are in 2.1.2 and 2.5.3, but not 2.2.z

**Comment by evgeni** [comment]
I've talked to Ina and she'll backport that for us ♥

**Comment by ehelms** [comment]
This is currently failing due to https://community.theforeman.org/t/missing-pulp-3-workers-on-upgrade-to-katello-4-0/24060

**Comment by ehelms** [comment]
Since 4.0 is being a problem child, we discussed putting a check in place to only run the intermediate tests if the final version is 4.1+.

File: pipelines/upgrade_pipeline.yml [hunk 3]
```diff
@@ -6,6 +6,7 @@
 - import_playbook: upgrade/05-server_to_intermediate.yaml
 - import_playbook: upgrade/06-proxy_to_intermediate.yaml
 - import_playbook: upgrade/07-server_to_final.yaml
-- import_playbook: upgrade/08-proxy_to_final.yaml
-- import_playbook: upgrade/09-tests.yaml
-- import_playbook: upgrade/10-smoker.yaml
+- import_playbook: upgrade/08-tests.yaml
```
Comment (ehelms) [inline pipelines/upgrade_pipeline.yml:9]:
```suggestion
 - import_playbook: upgrade/08-tests.yaml
+  when:
+    - forklift_upgrade_version_final == 'nightly' or forklift_upgrade_version_final is version('4.1', '>=')
```

**Comment by ehelms** [comment]
Needs a rebase as well. Current testing of nightly is blocked due to a bats tests failure with Puppet extraction merged.

File: pipelines/upgrade_pipeline.yml (hunk 3 above)
Comment (evgeni) [reply pipelines/upgrade_pipeline.yml:9]: this wouldn't have worked if run by jenkins (which doesn't use this file)

I pushed a version that should work everywhere

File: pipelines/upgrade/08-tests.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      scenario: "{{ pipeline_type }}"
+      scenario_os: "{{ pipeline_os }}"
+      scenario_version: "{{ pipeline_version }}"
+      upgrade: True
+    - role: forklift_versions
+      scenario: "{{ pipeline_type }}"
+      scenario_os: "{{ pipeline_os }}"
+      scenario_version: "{{ forklift_upgrade_version_final }}"
+    - role: foreman_client_repositories
+      when: pipeline_type != 'foreman'
```
Comment (ehelms) [inline pipelines/upgrade/08-tests.yaml:25]: Should we gate this role also on the condition at line 28?

**Review by ehelms — APPROVED** [review]
Minor comment, I moved this out of draft as it passed testing for me on 4.1 and nightly. I would be happy to have it merged with a squash and addressing comment.

File: pipelines/upgrade/08-tests.yaml (hunk 4 above)
Comment (evgeni) [reply pipelines/upgrade/08-tests.yaml:25]: It doesn't really hurt to run the repositories role here, and it doesn't take much time either. 

Or do you have other reasons to skip it here?

Comment (ehelms) [reply pipelines/upgrade/08-tests.yaml:25]: Just time saving -- every second counts when it takes eleventy-hours to run

**Pull request merged**